File: packages/typescript/src/telemetry/logger.ts
import type { Logger, LogLevel, LogContext } from './types.js';
import { LOG_LEVEL_PRIORITY } from './types.js';

/**
 * Patterns for sensitive data that should be redacted
 */
const REDACT_PATTERNS = [
    /api[_-]?key/i,
    /authorization/i,
    /x-luna-api-key/i,
    /password/i,
    /secret/i,
    /token/i,
    /bearer/i,
];

/**
 * Console-based logger with redaction support
 */
export class ConsoleLogger implements Logger {
    private readonly levelPriority: number;

    constructor(level: LogLevel = 'info') {
        this.levelPriority = LOG_LEVEL_PRIORITY[level];
    }

    error(message: string, context?: LogContext): void {
        this.log('error', message, context);
    }

    warn(message: string, context?: LogContext): void {
        this.log('warn', message, context);
    }

    info(message: string, context?: LogContext): void {
        this.log('info', message, context);
    }

    debug(message: string, context?: LogContext): void {
        this.log('debug', message, context);
    }

    trace(message: string, context?: LogContext): void {
        this.log('trace', message, context);
    }

    private log(level: LogLevel, message: string, context?: LogContext): void {
        // Check if this level should be logged
        if (LOG_LEVEL_PRIORITY[level] < this.levelPriority) {
            return;
        }

        const timestamp = new Date().toISOString();
        const sanitizedContext = context ? this.sanitize(context) : undefined;

        const logEntry = {
            timestamp,
            level: level.toUpperCase(),
            message,
            sdk: 'luna-sdk',
            version: '1.0.0',
            language: 'typescript',
            ...(sanitizedContext && { context: sanitizedContext }),
        };

        const output = JSON.stringify(logEntry);

        switch (level) {
            case 'error':
                console.error(output);
                break;
            case 'warn':
                console.warn(output);
                break;
            case 'trace':
            case 'debug':
                console.debug(output);
                break;
            default:
                console.log(output);
        }
    }

    /**
     * Sanitize context by redacting sensitive values
     */
    private sanitize(obj: LogContext): LogContext {
        const result: LogContext = {};

        for (const [key, value] of Object.entries(obj)) {
            if (this.isSensitiveKey(key)) {
                result[key] = '[REDACTED]';
            } else if (value && typeof value === 'object' && !Array.isArray(value)) {
                result[key] = this.sanitize(value as LogContext);
            } else {
                result[key] = value;
            }
        }

        return result;
    }

    /**
     * Check if a key contains sensitive data
     */
    private isSensitiveKey(key: string): boolean {
        return REDACT_PATTERNS.some((pattern) => pattern.test(key));
    }
}

File: packages/go/luna/telemetry/logger.go
// Package telemetry provides logging and metrics for the Luna SDK.
package telemetry

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"

	"time"
)

// LogLevel represents logging verbosity
type LogLevel int

const (
	LogLevelError LogLevel = 50
	LogLevelWarn  LogLevel = 40
	LogLevelInfo  LogLevel = 30
	LogLevelDebug LogLevel = 20
	LogLevelTrace LogLevel = 10
)

// Logger is the interface for SDK logging
type Logger interface {
	Error(message string, context map[string]interface{})
	Warn(message string, context map[string]interface{})
	Info(message string, context map[string]interface{})
	Debug(message string, context map[string]interface{})
	Trace(message string, context map[string]interface{})
}

// ConsoleLogger logs to stdout/stderr with JSON formatting
type ConsoleLogger struct {
	level    LogLevel
	redactRe []*regexp.Regexp
}

// NewConsoleLogger creates a new console logger
func NewConsoleLogger(level LogLevel) *ConsoleLogger {
	return &ConsoleLogger{
		level: level,
		redactRe: []*regexp.Regexp{
			regexp.MustCompile(`(?i)api[_-]?key`),
			regexp.MustCompile(`(?i)authorization`),
			regexp.MustCompile(`(?i)x-luna-api-key`),
			regexp.MustCompile(`(?i)password`),
			regexp.MustCompile(`(?i)secret`),
			regexp.MustCompile(`(?i)token`),
			regexp.MustCompile(`(?i)bearer`),
			// POPIA / SA Specific
			regexp.MustCompile(`(?i)id[_-]?number`),
			regexp.MustCompile(`(?i)tax[_-]?ref`),
			regexp.MustCompile(`(?i)registration[_-]?number`),
			regexp.MustCompile(`(?i)account[_-]?number`),
			regexp.MustCompile(`(?i)cvv`),
			regexp.MustCompile(`(?i)pan`),
		},
	}
}

func (l *ConsoleLogger) Error(message string, context map[string]interface{}) {
	l.log(LogLevelError, "ERROR", message, context)
}

func (l *ConsoleLogger) Warn(message string, context map[string]interface{}) {
	l.log(LogLevelWarn, "WARN", message, context)
}

func (l *ConsoleLogger) Info(message string, context map[string]interface{}) {
	l.log(LogLevelInfo, "INFO", message, context)
}

func (l *ConsoleLogger) Debug(message string, context map[string]interface{}) {
	l.log(LogLevelDebug, "DEBUG", message, context)
}

func (l *ConsoleLogger) Trace(message string, context map[string]interface{}) {
	l.log(LogLevelTrace, "TRACE", message, context)
}

func (l *ConsoleLogger) log(level LogLevel, levelStr, message string, context map[string]interface{}) {
	if level < l.level {
		return
	}

	entry := map[string]interface{}{
		"timestamp": time.Now().UTC().Format(time.RFC3339),
		"level":     levelStr,
		"message":   message,
		"sdk":       "luna-sdk",
		"version":   "1.0.0",
		"language":  "go",
	}

	if context != nil {
		entry["context"] = l.sanitize(context)
	}

	output, _ := json.Marshal(entry)

	if level >= LogLevelError {
		fmt.Fprintln(os.Stderr, string(output))
	} else {
		fmt.Println(string(output))
	}
}

func (l *ConsoleLogger) sanitize(obj map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})

	for key, value := range obj {
		if l.isSensitiveKey(key) {
			result[key] = "[REDACTED]"
		} else if nested, ok := value.(map[string]interface{}); ok {
			result[key] = l.sanitize(nested)
		} else {
			result[key] = value
		}
	}

	return result
}

func (l *ConsoleLogger) isSensitiveKey(key string) bool {
	for _, re := range l.redactRe {
		if re.MatchString(key) {
			return true
		}
	}
	return false
}

var _ Logger = (*ConsoleLogger)(nil)
